Feature/sb drawer changes

emclaug2 opened the pull request:
<!-- Include a bulleted list summarizing the main changes you have made in this PR -->
#### Changes proposed in this Pull Request:
- Update the Drawer API as follows:
- Use a 4dp shadow on the drawer by default
- Add a sideBorder prop (boolean), off by default
- Font weight 400 for all inactive nav items
- NaviItem dividers off by default
- Square active background by default
- Font weight 600 for active item (title)
- Use primary[500] color with 5% opacity for active highlight (primary 500 is a theme change)
- Use Overline variant for NavGroup title by default
- DrawerFooter divider prop should be true by default
- Add an option to use condensed nav items (40px tall, with 8px vertical margins)
- DrawerFooter/DrawerSubheader add a boolean prop to enable/disable hiding the content on collapse (default to hide)
- For nested menu items, if a nested item is active, all parent items up the tree should have fontWeight 600 for the title.

<!-- Include screenshots if they will help illustrate the changes in this PR -->
#### Screenshots:
![image](https://user-images.githubusercontent.com/6538289/97492196-97783580-1939-11eb-9c66-295164de8018.png)

emclaug2 commented:
TODO: 

- Update Docs with class name changes / default settings
- Add more unit tests
- Add new storybook knobs
- Drawer side border shadow & prop

joebochill commented:
@huayunh  @Horchen154 What do you guys think about making the "semi bold title for all parent elements of active element" configurable/opt-out-able? We never user tested that and I have a feeling I'm not the only one who will find it confusing/busy.

Horchen154 commented:
I dont have a problem with that as long as there is some other type of style applied to show that the parent element is active.

huayunh commented:
I am OK with it but I will probably prefer it as a style override instead of a prop/attr `pxb-drawer-active-item-parents`.

Here are two extra examples

Material.io
![image](https://user-images.githubusercontent.com/8997218/97601310-6b63bf80-19e0-11eb-87f5-5733811f4632.png)

Material UI
![image](https://user-images.githubusercontent.com/8997218/97601388-80405300-19e0-11eb-9a0e-103f86c18e99.png)

(I know they don't have subtitles... maybe that's why you feel busy)

emclaug2 commented:
I've removed the permanent divider under the DrawerHeader since it was causing the Header to be 1px taller than a traditional MatToolbar.  I've also removed the manual 64px specification from our DrawerHeader; The MatToolbar is 64px by default, and narrows for 56px for mobile viewports.  

Should I add a divider prop to the DrawerHeader, similar to DrawerFooter?

emclaug2 commented:
> Should I add a divider prop to the DrawerHeader, similar to DrawerFooter?

`divider` prop added to DrawerHeader with default false.

emclaug2 commented:
Latest Drawer w/ theme changes here: https://pxblue-angular-library.web.app/

huayunh commented:
Remember to rotate the chevron / drop down arrows when the nested items are expanded

emclaug2 commented:
> Remember to rotate the chevron / drop down arrows when the nested items are expanded

Great catch, I'm going to fix this today.

joebochill commented:
> I am OK with it but I will probably prefer it as a style override instead of a prop/attr `pxb-drawer-active-item-parents`.
> 
> Here are two extra examples
> 
> Material.io
> ![image](https://user-images.githubusercontent.com/8997218/97601310-6b63bf80-19e0-11eb-87f5-5733811f4632.png)
> 
> Material UI
> ![image](https://user-images.githubusercontent.com/8997218/97601388-80405300-19e0-11eb-9a0e-103f86c18e99.png)
> 
> (I know they don't have subtitles... maybe that's why you feel busy)

Notice how Material UI does not do what we're proposing. Their parent elements are always bold and they have a single selected style (blue + medium).

Material.io doesn't seem to be as confusing because they are using simple single-line text elements in the drawer. I think because we are using ILIs, it's more confusing because the style is only changing for one part of it (title). 

I would prefer to actually user test this and see if people can make sense of it before we roll it out. If they get it and don't have any issues, then I will happily concede.

Since this was a stretch goal on the story, is there a way we can comment it out or make it quickly toggleable (for the purposes of testing it with people and not holding up the other agreed-upon updates that are in this PR?

joebochill reviewed (commented):
Changelog? All these class name changes seems like a new major version to me, yeah?

emclaug2 commented on demos/storybook/firebase.json:
```diff
@@ -0,0 +1,6 @@
+{
```
Good catch, this is removed now.

emclaug2 commented on components/src/core/drawer/drawer-footer/drawer-footer.component.ts:
```diff
@@ -19,6 +22,7 @@ import { StateListener } from '../state-listener.component';
 })
 export class DrawerFooterComponent extends StateListener {
     @Input() divider = true;
+    @Input() hideOnCollapse = true;
```
Updated the prop name

emclaug2 commented on components/src/core/drawer/drawer-body/nav-item/drawer-nav-item.component.scss:
```diff
@@ -33,6 +36,13 @@ pxb-drawer-nav-item {
+    &.pxb-drawer-nav-item-compact .pxb-info-list-item-content {
+        &.pxb-info-list-item-dense,
+        .mat-list-item-content {
+            height: 40px !important;
```
Yeah this is tricky; if users don't like the `compact` height they can just manually override the height with their own css.

joebochill commented on demos/storybook/.gitignore:
```diff
@@ -0,0 +1,65 @@
+# Logs
```
I actually don't know if this file will do anything since we have a gitignore file at the root of the project. What's the intent here?

emclaug2 replied:
I'm not entirely sure why this got here; maybe this was generated when I locally made a firebase project in this folder.  I've removed this file.

joebochill commented on docs/Drawer.md:
```diff
@@ -23,10 +23,11 @@ Parent element (`<pxb-drawer>`) attributes:
+| ---------- | -------------------------------------- | --------- | -------- | ------- |
+| condensed  | Skinny view for `rail` variant         | `boolean` | no       | false   |
+| open       | State for the drawer                   | `boolean` | yes      |         |
+| sideBorder | Toggle a side border                   | `boolean` | no       | false   |
```
```suggestion
| sideBorder | Toggle a side border instead of shadow | `boolean` | no       | false   |
```

joebochill commented on docs/Drawer.md:
```diff
@@ -297,19 +307,20 @@ Parent element (`<pxb-drawer-nav-item>`) attributes:
+| @input                    | Description                           | Type                 | Required | Default  |
+| ------------------------- | ------------------------------------- | -------------------- | -------- | -------- |
+| activeItemBackgroundShape | Sets the active item background shape | `'round' | 'square'` | no       | 'square' |
+| compact                   | Sets whether to use narrow nav items  | `boolean`            | no       | false    |
```
```suggestion
| compact                   | Sets whether to use compact nav items  | `boolean`            | no       | false    |
```

joebochill replied:
in my mind narrow = width

joebochill commented on components/src/core/drawer/drawer-body/nav-item/drawer-nav-item.component.scss:
```diff
@@ -33,6 +36,13 @@ pxb-drawer-nav-item {
+    &.pxb-drawer-nav-item-compact .pxb-info-list-item-content {
+        &.pxb-info-list-item-dense,
+        .mat-list-item-content {
+            height: 40px !important;
```
@huayunh correct me if I'm wrong, but I believe the compact version is also supposed to have a margin between drawer items.

emclaug2 replied:
the compact height is 40px total.

Also I've added in a conditional that removes the subtitle from a navItem if the compact density is used.

joebochill commented:
Looking really good. Can you make sure we have a story on the backlog (or add a bullet to the testing story to create it) to come back to this once we finish the testing on the active parent styles?

joebochill commented:
The compact version should be 40px tall with 8px top and bottom margin:
![image](https://user-images.githubusercontent.com/29152776/98237747-639da100-1f33-11eb-8f84-42418611218f.png)

@huayunh this variation is not in the DSM. Do we need to make a story to add it?

huayunh commented:
This can be easily fixed while we rebuild part of drawer using Figma’s variants

emclaug2 commented:
@joebochill @huayunh 
I've added in the 8px margin between compact nav-items.  It can be be seen [here](https://pxblue-angular-library.web.app/?path=/story/components-drawer--with-full-config&knob-title_DrawerHeader=PX Blue Drawer&knob-subtitle_DrawerHeader=with full config&knob-Show Background Image_DrawerHeader=true&knob-divider_DrawerHeader=true&knob-NavGroup 1 title_DrawerNavGroup=Group 1&knob-NavGroup 2 title_DrawerNavGroup=Group 2&knob-divider_DrawerNavGroup=true&knob-Add Spacer_DrawerNavGroup=&knob-Group 1 Items_DrawerNavGroup=3&knob-Group 2 Items_DrawerNavGroup=3&knob-enableSelectionHierarchy_DrawerNavItem=&knob-chevron_DrawerNavItem=&knob-hidePadding_DrawerNavItem=true&knob-hidePadding (nested)_DrawerNavItem=&knob-divider_DrawerNavItem=&knob-Show Icon_DrawerNavItem=true&knob-compact_DrawerNavItem=true&knob-Custom Expand/Collapse Icons_DrawerNavItem=&knob-activeItemBackgroundShape_DrawerNavItem=square&knob-Show Footer_DrawerFooter=true&knob-divider_DrawerFooter=true&knob-hideContentOnCollapse_DrawerFooter=true).

I'm not sure if there should be special treatment for the nav items with children.  Right now there is a 8px margin applied with 0 exceptions.

huayunh commented:
![image](https://user-images.githubusercontent.com/8997218/98389369-f288e700-2021-11eb-9e7d-aa7a4c8c9a5b.png)
Blue 500 for highlighted items?

huayunh commented:
NavGroup header Divider:

in the storybook, set the top divider on and set the bottom divider off by default

emclaug2 commented:
> ![image](https://user-images.githubusercontent.com/8997218/98389369-f288e700-2021-11eb-9e7d-aa7a4c8c9a5b.png)
> Blue 500 for highlighted items?

This is a theme change; the firebase app wasn't use latest theme.
Check again in incognito; should be working now. 

The navGroup header this is this something that users can already do; can you make a new issue for this so this PR can get merged?

emclaug2 commented:
> > ![image](https://user-images.githubusercontent.com/8997218/98389369-f288e700-2021-11eb-9e7d-aa7a4c8c9a5b.png)
> > Blue 500 for highlighted items?
> 
> This is a theme change; the firebase app wasn't use latest theme.
> Check again in incognito; should be working now.
> 
> Regarding navGroup header top/bottom dividers - this is this something that users can already do; can you make a new issue for adding @inputs for top/bottom dividers so this PR can get merged?
